temporal-sdk/src/test/java/io/temporal/internal/worker/WorkflowWorkerTest.java

```diff
@@ -0,0 +1,196 @@
+        .thenReturn(pollResponse)
+        .thenReturn(pollResponse)
+        .thenAnswer(
+            (Answer<PollWorkflowTaskQueueResponse>)
+                invocation -> {
+                  pollTaskQueueLatch.countDown();
+                  return pollResponse;
+                })
+        .thenAnswer(
+            (Answer<PollWorkflowTaskQueueResponse>)
+                invocation -> {
+                  blockPollTaskQueueLatch.await();
+                  return null;
+                });
+
+    CountDownLatch handleTaskLatch = new CountDownLatch(1);
+    when(taskHandler.handleWorkflowTask(any(PollWorkflowTaskQueueResponse.class)))
+        .thenAnswer(
+            (Answer<WorkflowTaskHandler.Result>)
+                invocation -> {
+                  // Slightly larger than the lock timeout hard coded in WorkflowWorker
+                  handleTaskLatch.countDown();
+                  Thread.sleep(6000);
+
+                  return new WorkflowTaskHandler.Result(
+                      WORKFLOW_TYPE,
+                      RespondWorkflowTaskCompletedRequest.newBuilder().build(),
+                      null,
+                      null,
+                      null,
+                      false,
+                      (id) -> {
+                        // verify the lock is still being held
+                        assertEquals(runLockManager.totalLocks(), 1);
+                      });
+                });
+
+    // Mock the server responding to a workflow task complete with another workflow task
+    CountDownLatch respondTaskLatch = new CountDownLatch(1);
+    when(blockingStub.respondWorkflowTaskCompleted(any(RespondWorkflowTaskCompletedRequest.class)))
+        .thenAnswer(
+            (Answer<RespondWorkflowTaskCompletedResponse>)
+                invocation -> {
+                  // verify the lock is still being held
+                  assertEquals(runLockManager.totalLocks(), 1);
+                  return RespondWorkflowTaskCompletedResponse.newBuilder()
+                      .setWorkflowTask(pollResponse)
+                      .build();
+                })
+        .thenAnswer(
+            (Answer<RespondWorkflowTaskCompletedResponse>)
+                invocation -> {
+                  // verify the lock is still being held
+                  assertEquals(runLockManager.totalLocks(), 1);
+                  respondTaskLatch.countDown();
+                  return RespondWorkflowTaskCompletedResponse.newBuilder().build();
+                });
+
+    assertTrue(worker.start());
+    // Wait until we have got all the polls
+    pollTaskQueueLatch.await();
+    // Wait until the worker handles at least one WFT
+    handleTaskLatch.await();
+    // Sleep to allow metrics to be published
+    Thread.sleep(100);
+    // Since all polls have the same runID only one should get through, the other two should be
+    // blocked
+    assertEquals(runLockManager.totalLocks(), 1);
```

Comment:
The comments are good - you might consider including them in the assert call as Strings so that they show up in the test failure output as well.